### ISSUE TITLE
[Feature request] Enhance large files with syntax highlighting performance

### ISSUE THREAD

**Issue by softmgr** [notepad-plus-plus/notepad-plus-plus#15952]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

**Steps to Reproduce Performance Issues in Notepad++:**
1. Open a large file (e.g., an HTML file larger than 100MB) in Notepad++ with syntax highlighting enabled.
2. Press Ctrl + End. At this point, Notepad++ becomes unresponsive and takes a significant amount of time to recover.
3. Reopen the large file, press Ctrl + G, input a very large line number, and press Enter. Again, Notepad++ becomes unresponsive and takes a long time to respond.
4. Move the cursor to the end of the large file. Then, create a new file or open another file. Finally, switch back to the tab containing the large file. Notepad++ takes a significant amount of time to respond.

**Fix Description:**
@donho I have optimized the behavior for the above scenarios. The attached source code includes my modifications, based on the latest master branch as of today.

[notepad-plus-plus-master[20241215].zip](https://github.com/user-attachments/files/18140068/notepad-plus-plus-master.20241215.zip)

### Describe the solution you'd like.

I have optimized the behavior for the above scenarios. 

### Debug Information

```shell
Notepad++ v8.7.4   (64-bit)
Build time : Dec  4 2024 - 23:50:05
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
Periodic Backup : ON
Placeholders : OFF
DirectWrite : OFF
Multi-instance Mode : monoInst
File Status Auto-Detection : cdEnabledNew (for current file/tab only)
Dark Mode : ON
OS Name : Windows 10 Enterprise (64-bit)
OS Version : 22H2
OS Build : 19045.5131
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3.1)
```

### Anything else?

**Relevant links:**
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/15653
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/15926

**Comment by donho** [comment]
@softmgr 
> Press Ctrl + End. At this point, Notepad++ becomes unresponsive and takes a significant amount of time to recover.

It should be fixed by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/de9ffd2ea8507d033f7f111d8b48762f7d3b9436 which is in the current master.
However, I can still reproduce on my PC (Unfortunately, I have no other PC for checking currently)
Could you please check it and confirm me if it's fixed for you?

**Comment by softmgr** [comment]
> @softmgr 
> 
> > Press Ctrl + End. At this point, Notepad++ becomes unresponsive and takes a significant amount of time to recover.
> 
> 
> 
> It should be fixed by https://github.com/notepad-plus-plus/notepad-plus-plus/commit/de9ffd2ea8507d033f7f111d8b48762f7d3b9436 which is in the current master.
> 
> However, I can still reproduce on my PC (Unfortunately, I have no other PC for checking currently)
> 
> Could you please check it and confirm me if it's fixed for you?

I compiled the latest master branch, and the performance for opening and switching files is generally as expected.
Additionally, the code `runMarkers(true, 0, true, false);` in the function `activateBuffer()` has a significant impact on the performance of switching tabs, causing a noticeable lag. Perhaps this function needs some optimization?
Using `execute(SCI_MARKERNEXT, searchStart, static_cast<LPARAM>(1 << MARK_HIDELINESBEGIN));` to replace `auto state = execute(SCI_MARKERGET, i);` might help.

When execute(SCI_MARKERNEXT, ...) == -1, it means there are no more markers, and the function can exit directly without iterating through all the lines. What do you think?

**Comment by donho** [comment]
@softmgr 
How about pressing Ctrl + End? Does it do as expected?

> Additionally, the code runMarkers(true, 0, true, false); in the function activateBuffer() has a significant impact on the performance of switching tabs, causing a noticeable lag. Perhaps this function needs some optimization?
Using execute(SCI_MARKERNEXT, searchStart, static_cast<LPARAM>(1 << MARK_HIDELINESBEGIN)); to replace auto state = execute(SCI_MARKERGET, i); might help.

I'll check it.

**Comment by softmgr** [comment]
> @softmgr 
> 
> How about pressing Ctrl + End? Does it do as expected?
> 

Yes, now pressing Ctrl + End positions the cursor quickly at the end of the file, and switching tabs is also very fast. However, the unresolved issue is that switching tabs while using rectangular selection still has performance problems. This issue requires modifying the code in the Editor.cxx file of the Scintilla component.

**Comment by donho** [comment]
> Yes, now pressing Ctrl + End positions the cursor quickly at the end of the file

Thank you for confirming it. I ask it because on laptop it doesn't work as expected. It could be due to an update of Windows 11

**Comment by softmgr** [comment]
> > Yes, now pressing Ctrl + End positions the cursor quickly at the end of the file
> 
> 
> 
> Thank you for confirming it. I ask it because on laptop it doesn't work as expected. It could be due to an update of Windows 11

How does it perform on Windows 11?

**Comment by donho** [comment]
Pressing Ctrl + End quickly positions the cursor at the end of the file, without syntax highlighting, and few seconds later syntax highlighting appears - It was the old behaviour before the updating of Windows on my laptop.

Now after launching by opening a large file and pressing Ctrl + End - I have to wait several seconds to position the cursor at the end of the file, but with syntax highlighting.

My OS information is:
```
OS Name : Windows 11 Home (64-bit)
OS Version : 24H2
OS Build : 26100.2605
```

**Comment by softmgr** [comment]
> Pressing Ctrl + End quickly positions the cursor at the end of the file, without syntax highlighting, and few seconds later syntax highlighting appears - It was the old behaviour before the updating of Windows on my laptop.
> 
> 
> 
> Now after launching by opening a large file and pressing Ctrl + End - I have to wait several seconds to position the cursor at the end of the file, but with syntax highlighting.
> 
> 
> 
> My OS information is:
> 
> ```
> 
> OS Name : Windows 11 Home (64-bit)
> 
> OS Version : 24H2
> 
> OS Build : 26100.2605
> 
> ```



What about disabling ClearType and DirectWrite in the preferences?

**Comment by donho** [comment]
- ClearType on Windows: OFF
- DirectWrite in Notepad++: OFF

The same result (wait several seconds to position the cursor at the end of the file, with syntax highlighting) after restarting the Windows. :(

**Comment by softmgr** [comment]
> - ClearType on Windows: OFF
> 
> - DirectWrite in Notepad++: OFF
> 
> 
> 
> The same result (wait several seconds to position the cursor at the end of the file, with syntax highlighting) after restarting the Windows. :(
> 
> 
> 

Let me upgrade the Windows 11 virtual machine to the latest version and test it.

**Comment by donho** [comment]
>>Additionally, the code runMarkers(true, 0, true, false); in the function activateBuffer() has a significant impact on the performance of switching tabs, causing a noticeable lag. Perhaps this function needs some optimization?
Using execute(SCI_MARKERNEXT, searchStart, static_cast(1 << MARK_HIDELINESBEGIN)); to replace auto state = execute(SCI_MARKERGET, i); might help.

>I'll check it.

@softmgr 
Please check #16010 which should improve `activateBuffer`.

**Comment by softmgr** [comment]
> - ClearType on Windows: OFF
> 
> - DirectWrite in Notepad++: OFF
> 
> 
> 
> The same result (wait several seconds to position the cursor at the end of the file, with syntax highlighting) after restarting the Windows. :(
> 
> 
> 

I have tested running Notepad++ on the latest version of Windows 11 24H2, opening and switching large files is very fast. The performance is completely consistent with that on Windows 10.

During testing, both ClearType and DirectWrite were enabled, specifically:
 • Settings -> Preferences -> Editing 1 -> Enable smooth font: Enabled
 • Settings -> Preferences -> MISC. -> Use DirectWrite: Enabled

Here is the operating system information used in my tests:
 • OS Name: Windows 11 IoT Enterprise LTSC 2024 (64-bit)
 • OS Version: 24H2
 • OS Build: 26100.2605

**Comment by donho** [comment]
@softmgr 
Thank you for testing. I guess it could be an isolated issue from my laptop.

**Comment by donho** [comment]
@softmgr 
> I believe this might be caused by an issue with the implementation. I recommend reviewing the language-switching code and optimizing the method of invocation.

Will do.

> This problem is unrelated to the Windows version. When initially loading a large file, if Word Wrap is enabled, the loading process will be very slow. However, if this option is disabled, large files can be loaded quickly.

Yes indeed. I should have thought about Word Wrap, since it causes a lot of performance issue.
Thank you for your investigation!

**Comment by donho** [comment]
Unfortunately the code provided causes the regression (#16309). Part of code were reversed. See #16312

**Comment by softmgr** [comment]
@donho 
The performance issue related to automatic line wrapping in large files has been resolved:

The root cause was in Scintilla’s source code—when opening a file, syntax highlighting and automatic line wrapping were being executed simultaneously, which caused the application to become unresponsive.

To address this, I enabled idle-time syntax highlighting during editor initialization using:
```
execute(SCI_SETIDLESTYLING, SC_IDLESTYLING_ALL, 0);
```
This setting causes Scintilla to perform syntax highlighting incrementally using `SetTimer`, rendering portions of text every 10 milliseconds until completion.

However, during this process, if automatic line wrapping is also enabled, Scintilla calls:
```
WrapLines(WrapScope::wsVisible);
```
This means it immediately wraps all visible lines. When pressing Ctrl+End, the visible area becomes the entire document. For large files, this results in the program freezing.

The solution requires modifying Scintilla’s source code:
 1. Replace all instances of `WrapLines(WrapScope::wsAll);` with `WrapLines(WrapScope::wsIdle);` in Editor.cxx.

 2. Modify the Paint() function in Editor.cxx as follows:
```
void Editor::Paint(Surface *surfaceWindow, PRectangle rcArea) {
	// ...
	if (NotifyUpdateUI()) {
		RefreshStyleData();
	}

	const bool idlerRunning = idler.state && needIdleStyling;
	// Wrap the visible lines if needed.
	if ((!idlerRunning && WrapLines(WrapScope::wsVisible)) || (idlerRunning)) {  // if (WrapLines(WrapScope::wsVisible)) {
		// The wrapping process has changed the height of some lines so
		// abandon this paint for a complete repaint.
		if (AbandonPaint()) {
			return;
		}
	}

	// ...
}
```

With these changes, opening and switching between large files is now smooth and responsive, regardless of their size.
